### ISSUE TITLE
[Feature Request] Add API for manually setting logits.

### ISSUE THREAD

**Issue by shaunabanana** [ggerganov/llama.cpp#1024]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [X] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

Add a public API (e.g., `llama_set_logits(ctx, logits)`) that allows to manually set the logits. This would open the door to bridging llama.cpp with use cases like [FUDGE](https://arxiv.org/abs/2104.05218) ([code](https://github.com/yangkevin2/naacl-2021-fudge-controlled-generation)) and [Grounded Decoding](https://grounded-decoding.github.io/), which guides the generation process by manipulating the probabilities without the need for fine-tuning.

# Current Behavior

I understand that there is currently an API function for getting the current logits (`llama_get_logits(ctx)`), but no way to modify the logits through a public API.

**Comment by SlyEcho** [comment]
The function `llama_get_logits()` returns the pointer to the logits array in llama.cpp's state. To modify the logits, you just have to write to the array. You can an example in main.cpp itself:

```c++
auto logits = llama_get_logits(ctx);
if (params.ignore_eos) {
    logits[llama_token_eos()] = 0;
}
```

Still, maybe not the best API, better would be something like this which makes a copy:

```c
// length from llama_n_vocab()
LLAMA_API int llama_get_logits(struct llama_context * ctx, float * logits);
LLAMA_API int llama_set_logits(struct llama_context * ctx, const float * logits);
```

**Comment by shaunabanana** [comment]
Thank you for pointing this out! I'll see if I can work out something :)
